### PR TITLE
Adding and using the CacheManager

### LINKED ISSUE #343: Unable to manage branches after re-opening GitHub window
### Description
I'm unable to view/create/delete my branches in a couple different scenarios. The workaround I've found when I get into these scenarios is to restart Unity completely, but that's not ideal.

#### Scenario 1
1. Create new project, initialize repo, publish
2. In GitHub window, go to `Branches` tab
3. See if you can view/create/delete branches

#### Scenario 2
1. Start Unity
2. If GitHub window was open before, it will be open again and you can manage branches fine.
3. Close GitHub window
4. Re-open GitHub window
5. See if you can view/create/delete branches

**Expected behavior for both scenarios**:
Can manage my branches
screen shot 2017-09-21 at 10 35 57 am

**Actual behavior**:
branch situation

### Additional Information
This behaves the same in both Windows and Mac with the latest build `0.21.1`. I don't see any helpful information in the logs :(

I only ended up trying to close and open the tab because I'm not able to see my newly created branch without trying to "jiggle the handle" as Stanley would say.

### LINKED ISSUE #344: Master branch does not appear in branches list after initialization
1. Initialize Repo
1. Goto Branches View
1. Observe master branch is not present in "Local Branches"
1. Close & Reopen Unity
1. Goto Branches View
1. Observe master branch is present in "Local Branches"

### LINKED ISSUE #357: Changed View not rendering tree view correctly after a script is compiled
### Description

In the UnityExtention project, the tree disappears after compiling a script.

### Steps to Reproduce

1. Open UnityExtention project
2. GitHub -> Show Window
3. Select the Changed tab
4. Open a script file and trigger a compile by modifying something and then saving
5. Go back to Unity

**Expected behavior:**
The Changed view should update the Tree View and display the changes

**Actual behavior:**
The Tree View shows the text "No changes found"

**Reproduces how often:**
Always (5.6.3p1)

### Additional Information
Switching to the History tab and back fixes it

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Fixes #357 
Fixes #344 
Fixes #343

Depends on:
- [x] #384 
- [x] #391 
- [x] #392
- [x] #393
- [x] #394
- [x] #396 
- [x] #397
- [x] #399
- [x] #389

### CODE REVIEW

File: src/GitHub.Api/Git/RepositoryManager.cs [hunk 1]
```diff
@@ -12,22 +12,21 @@ public interface IRepositoryManager : IDisposable
         event Action<IUser> OnGitUserLoaded;
         event Action<bool> OnIsBusyChanged;
         event Action<string> OnLocalBranchAdded;
-        event Action<Dictionary<string, ConfigBranch>> OnLocalBranchListUpdated;
+        event Action<IDictionary<string, ConfigBranch>> OnLocalBranchListUpdated;
```
Comment (shana) [inline src/GitHub.Api/Git/RepositoryManager.cs]: Why the change to the interface?

File: src/GitHub.Api/Git/RepositoryManager.cs [hunk 2]
```diff
@@ -457,11 +425,16 @@ private void UpdateCurrentBranchAndRemote(string head)
...
+                OnCurrentBranchUpdated?.Invoke(branch);
 
-            Logger.Trace("OnCurrentRemoteUpdated: {0}", remote.HasValue ? remote.Value.ToString() : "[NULL]");
-            OnCurrentRemoteUpdated?.Invoke(remote);
+                Logger.Trace("OnCurrentRemoteUpdated: {0}", remote.HasValue ? remote.Value.ToString() : "[NULL]");
+                OnCurrentRemoteUpdated?.Invoke(remote);
+            })
+            {
+                Affinity = TaskAffinity.UI
+            }.Start();
```
Comment (shana) [inline src/GitHub.Api/Git/RepositoryManager.cs]: No other methods here switch to the UI thread for updating data, so this one probably shouldn't either. Maybe let Repository handle threading, have that logic all in one place?

File: src/GitHub.Api/Git/Repository.cs [hunk 3]
```diff
@@ -10,49 +10,128 @@ namespace GitHub.Unity
...
             LocalPath = localPath;
-            this.User = new User();
+            User = new User();
+
+            cacheContainer = container;
+            cacheContainer.CacheInvalidated += CacheContainer_OnCacheInvalidated;
+            cacheContainer.CacheUpdated += CacheContainer_OnCacheUpdated;
+        }
+
+        private void CacheContainer_OnCacheInvalidated(CacheType cacheType)
```
Comment (shana) [inline src/GitHub.Api/Git/Repository.cs]: All these private methods should be after public methods, having a hard time reviewing this because of so many unnecessary method moves...

File: src/GitHub.Api/Git/Repository.cs [hunk 4]
```diff
@@ -10,49 +10,128 @@ namespace GitHub.Unity
...
+                case CacheType.BranchCache:
+                    HandleBranchCacheUpdatedEvent(cacheUpdateEvent);
+                    break;
+
+                case CacheType.GitLogCache:
+                    HandleGitLogCacheUpdatedEvent(cacheUpdateEvent);
+                    break;
+
+                case CacheType.GitStatusCache:
+                    HandleGitStatucCacheUpdatedEvent(cacheUpdateEvent);
```
Comment (shana) [inline src/GitHub.Api/Git/Repository.cs]: Statuc should be Status instead?

File: src/GitHub.Api/Git/Repository.cs [hunk 5]
```diff
@@ -61,11 +140,11 @@ public void Initialize(IRepositoryManager repositoryManager)
...
             repositoryManager.OnGitUserLoaded += user => User = user;
-        }
 
-        public void Refresh()
-        {
-            repositoryManager?.Refresh();
+            UpdateGitStatus();
+            UpdateGitLog();
+
+            new ActionTask(CancellationToken.None, UpdateLocks) { Affinity = TaskAffinity.UI }.Start();
```
Comment (shana) [inline src/GitHub.Api/Git/Repository.cs]: We shouldn't be doing these three actions on Initialize. Initialize is called every time the appdomain reloads, which means every time we enter playmode we'll be doing this. We only want to do this when we first load the project in order to update information, and afterwards only when the filesystem has changes that tell us we need to do these actions.

File: src/GitHub.Api/Git/Repository.cs [hunk 6]
```diff
@@ -107,7 +178,7 @@ public ITask Pull()
 
         public ITask Push()
         {
-            return repositoryManager.Push(CurrentRemote.Value.Name, CurrentBranch?.Name);
+            return repositoryManager.Push(CurrentRemote.Value.Name, CurrentBranch?.Name).Then(UpdateGitStatus);
```
Comment (shana) [inline src/GitHub.Api/Git/Repository.cs]: Should put the `Then` bit on a new line to make sure it's easy to read that `Push` has a significant side effect when it's done.

File: src/GitHub.Api/Git/Repository.cs [hunk 7]
```diff
@@ -120,21 +191,139 @@ public ITask Revert(string changeset)
             return repositoryManager.Revert(changeset);
         }
 
-        public ITask ListLocks()
+        public ITask RequestLock(string file)
         {
-            if (repositoryManager == null)
-                return new ActionTask(new NotReadyException().ToTask<bool>());
-            return repositoryManager.ListLocks(false);
+            return repositoryManager.LockFile(file).Then(UpdateLocks);
```
Comment (shana) [inline src/GitHub.Api/Git/Repository.cs]: For readability's sake, `Then` actions should be on their own line, they're pretty easy to miss otherwise

File: src/GitHub.Api/Git/Repository.cs [hunk 8]
```diff
@@ -120,21 +191,139 @@ public ITask Revert(string changeset)
...
         }
 
-        public ITask RequestLock(string file)
+        public ITask ReleaseLock(string file, bool force)
         {
-            return repositoryManager.LockFile(file);
+            return repositoryManager.UnlockFile(file, force).Then(UpdateLocks);
```
Comment (shana) [inline src/GitHub.Api/Git/Repository.cs]: Same as above wrt `Then`

File: src/GitHub.Api/Git/Repository.cs [hunk 9]
```diff
@@ -236,192 +461,191 @@ private void UpdateRepositoryInfo()
...
+        public GitRemote[] Remotes
         {
-            get
-            {
-                if (currentBranch != null)
-                {
-                    return GetLocalGitBranch(currentBranch.Value);
-                }
+            get { return cacheContainer.BranchCache.Remotes; }
+            set { cacheContainer.BranchCache.Remotes = value; }
```
Comment (shana) [inline src/GitHub.Api/Git/Repository.cs]: Shouldn't all these setters be private?

File: src/GitHub.Api/Git/Repository.cs [hunk 10]
```diff
@@ -236,192 +461,191 @@ private void UpdateRepositoryInfo()
...
-            CurrentRemote);
+        public event Action<CacheUpdateEvent> LogChanged;
+        public event Action<CacheUpdateEvent> StatusChanged;
+        public event Action<CacheUpdateEvent> CurrentBranchChanged;
+        public event Action<CacheUpdateEvent> CurrentRemoteChanged;
+        public event Action<CacheUpdateEvent> CurrentBranchAndRemoteChanged;
+        public event Action<CacheUpdateEvent> LocalBranchListChanged;
+        public event Action<CacheUpdateEvent> LocksChanged;
+        public event Action<CacheUpdateEvent> RemoteBranchListChanged;
+        public event Action<CacheUpdateEvent> LocalAndRemoteBranchListChanged;
```
Comment (shana) [inline src/GitHub.Api/Git/Repository.cs]: All these events should be up top

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs [hunk 11]
```diff
@@ -31,11 +31,17 @@ class Window : BaseWindow
...
+        [SerializeField] private string repoRemote;
         [SerializeField] private string repoBranch;
         [SerializeField] private string repoUrl;
         [SerializeField] private GUIContent repoBranchContent;
         [SerializeField] private GUIContent repoUrlContent;
 
+        [SerializeField] private CacheUpdateEvent lastCurrentBranchAndRemoteChangedEvent;
+        [NonSerialized] private bool currentBranchAndRemoteHasUpdate;
+
+        [NonSerialized] private bool hasRunMaybeUpdateDataWithRepository;
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs]: What's this flag for?

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs [hunk 12]
```diff
@@ -179,39 +182,76 @@ public override void Update()
...
             string updatedRepoRemote = null;
             string updatedRepoUrl = DefaultRepoUrl;
 
+            var shouldUpdateContentFields = false;
+
             if (Repository != null)
             {
-                var repositoryCurrentBranch = Repository.CurrentBranch;
-                updatedRepoBranch = repositoryCurrentBranch.HasValue ? repositoryCurrentBranch.Value.Name : null;
+                if(!hasRunMaybeUpdateDataWithRepository || currentBranchAndRemoteHasUpdate)
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs]: Missing space here

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs [hunk 13]
```diff
@@ -227,16 +267,28 @@ private void AttachHandlers(IRepository repository)
...
             if (repository == null)
                 return;
-            repository.OnRepositoryInfoChanged += RefreshOnMainThread;
-            repository.OnCurrentBranchUpdated += UpdateLog;
+            repository.CurrentBranchAndRemoteChanged += RepositoryOnCurrentBranchAndRemoteChanged;
+        }
+
+        private void RepositoryOnCurrentBranchAndRemoteChanged(CacheUpdateEvent cacheUpdateEvent)
+        {
+            if (!lastCurrentBranchAndRemoteChangedEvent.Equals(cacheUpdateEvent))
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs]: Is this ever false? Isn't Repository supposed to only raise events if the data has changed (and therefore the cacheUpdateEvent is always different here?

File: src/GitHub.Api/Git/RepositoryManager.cs (hunk 1 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/RepositoryManager.cs]: Hmm. It wasn't necessary. Rolling it back

Comment (StanleyGoldman) [reply src/GitHub.Api/Git/RepositoryManager.cs]: Yea, I originally made this change because I was trying to figure out serializations and
`Dictionary<string, Dictionary<string, ConfigBranch>>>` != `IDictionary<string, IDictionary<string, ConfigBranch>>>`

File: src/GitHub.Api/Git/Repository.cs (hunk 9 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/Repository.cs]: Correct...

File: src/GitHub.Api/Git/Repository.cs (hunk 6 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/Repository.cs]: 👍

File: src/GitHub.Api/Git/Repository.cs (hunk 7 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/Repository.cs]: 👍

File: src/GitHub.Api/Git/Repository.cs (hunk 8 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/Repository.cs]: 👍

File: src/GitHub.Api/Git/Repository.cs (hunk 10 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/Repository.cs]: 👍

File: src/GitHub.Api/Git/Repository.cs (hunk 4 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/Repository.cs]: 👍

File: src/GitHub.Api/Git/RepositoryManager.cs (hunk 2 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/RepositoryManager.cs]: 👍

File: src/GitHub.Api/Git/Repository.cs (hunk 3 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/Repository.cs]: 👍 

Apologies, I did a lot of hashing and rehashing

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs (hunk 12 above)
Comment (StanleyGoldman) [reply src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs]: 👍

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs (hunk 13 above)
Comment (StanleyGoldman) [reply src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs]: If you are looking at the `HistoryView` and switch to a new branch outside of Unity.
When you click on Unity, `Window` will update to the correct value.
Then switch to the `ChangesView` The changes view has an outdated value for `CurrentBranch`/`CurrentRemote`.
`ChangesView` will ask `Repository` to check it's cache timestamp.
`Repository` will see that the timestamp is out of date and re-fire the event.
in this scenario, line 275 will be false.

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs (hunk 11 above)
Comment (StanleyGoldman) [reply src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs]: When `Window` gets it's instance of `Repository` `currentBranchAndRemoteHasUpdate` was still false.
For some reason it's not having the effect I thought it did. It seems i can remove it.

File: src/GitHub.Api/Git/Repository.cs (hunk 5 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/Git/Repository.cs]: Where should I move it to?

Comment (shana) [reply src/GitHub.Api/Git/Repository.cs]: It probably needs to be something that happens when someone checks for the cache and the cache requests the data because it's either 1) too old, 2) has no data, 3) initialRun is true in ApplicationCache object. Maybe it makes sense for the cache objects to check that flag to see if they need to ask for new data when the views for ask for data? Maybe we should store a timestamp of when Unity first started up so the cache objects can check whether the cached data is older than data and request updated data when the views ask for data? Would that make sense?

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs (hunk 13 above)
Comment (shana) [reply src/UnityExtension/Assets/Editor/GitHub.Unity/UI/Window.cs]: Oooooh, yeah I see.

File: src/GitHub.Api/Git/Repository.cs (hunk 5 above)
Comment (shana) [reply src/GitHub.Api/Git/Repository.cs]: Tracking this change in https://github.com/github-for-unity/Unity/issues/409

**Comment by StanleyGoldman** [comment]
I was not sure what to do with data that was "derived" from cached data".
In the end I decided to lazily determine that data in the getters

https://github.com/github-for-unity/Unity/blob/08705a7250605a79fbc9a6fd1aa35df7144e2a13/src/GitHub.Api/Git/Repository.cs#L600-L639

I realize now that this function repeats the functionality to determine `Name` and `CloneUrl`.

https://github.com/github-for-unity/Unity/blob/08705a7250605a79fbc9a6fd1aa35df7144e2a13/src/GitHub.Api/Git/Repository.cs#L463-L477

**Comment by StanleyGoldman** [comment]
I have a dirty underbelly here...

`IRemoteConfigBranchDictionary` should no longer implement `IDictionary< ... , IDictionary< ... , ... >>`
That is left over from attempts to use `SerializableDictionary`. And as a result creates this mess here. 😓 

https://github.com/github-for-unity/Unity/blob/08705a7250605a79fbc9a6fd1aa35df7144e2a13/src/UnityExtension/Assets/Editor/GitHub.Unity/ApplicationCache.cs#L299-L397
